d3b-center/image-deid-etl
# Switch to a container-based development environment

## Overview

To produce a container image using Nix, I had to re-package the Python project in a way that satisfied Nix. I found myself falling down a rabbit hole, re-packaging the Python project using [Poetry](https://python-poetry.org/docs/) and trying to use [poetry2nix](https://github.com/nix-community/poetry2nix) to turn the project into a Nix derivation. We use a lot of Python dependencies that depend on C libraries (like `pyparsing`, `python-magic`, `pillow`, etc.), and I had to patch each one to fit the Nix way of making shared libraries available to packages. 

This research was OK because we intentionally baked in some time for exploring Nix, but I began to feel like my wheels were spinning, and I wouldn’t arrive at a workable solution in time for our early May deadline.

I decided to pivot to a more traditional approach using Docker and Docker Compose to manage the development environment and produce the container image. 

This approach feels a lot more straightforward. As a result, I can deliver faster, and we better minimize the divergence between development and production. However, we lose the reproducible builds and native performance of Nix. Also, we now have to deal with some [remote debugger trickery](https://www.jetbrains.com/help/pycharm/using-docker-compose-as-a-remote-interpreter.html#docker-compose-remote) with PyCharm.

I made the following changes to support this switch:

- Replace `flake.nix` with a `Dockerfile` for building a container image containing all of the CLI’s dependencies (e.g., the AWS CLI, dcm2niix, and the Flywheel CLI).
- Package the Python project using `setuptools` to [wrap the application](https://setuptools.pypa.io/en/latest/userguide/entry_point.html) in an `image-deid-etl` executable.
- Add service definitions for `docker-compose.yml` that mimic how we’d run the container on AWS.
- Wrap the `upload2fw` step in a fake user home directory so that the Flywheel CLI can consume credentials from the environment.
- Install the `psycopg2` and PostgreSQL client libraries and update the database schema to support PostgreSQL.

Resolves #23 
Resolves #9 

### Checklist


### Notes

At first, the ETL was noticeably slower than before. However, I updated to macOS Monterey 12.3 and tuned the following settings in Docker Desktop:

<details>

![image](https://user-images.githubusercontent.com/1774125/165814579-334fc3c1-6e5e-443d-820e-badb073723d8.png)

![image](https://user-images.githubusercontent.com/1774125/165814600-ad20d910-6e00-4f37-b323-54e17739593c.png)

</details>

While the ETL will never be as fast on macOS ran under Docker (since Docker for Mac is a Linux virtual machine), things _dramatically_ improved due to I/O gains with Docker's [experimental file system](https://www.docker.com/blog/speed-boost-achievement-unlocked-on-docker-desktop-4-6-for-mac/).

I compared the performance of this branch vs. `develop` processing a single Orthanc study:

<details>

**Docker:**

```console
$ rm -rf src/image_deid_etl/cbtn 
$ time docker-compose run --rm image-deid-etl image-deid-etl run f90cd746-10017fbb-44472a28-391f1a52-3186cf1a
. . .
docker-compose run --rm image-deid-etl image-deid-etl run   0.46s user 0.23s system 0% cpu 1:19.04 total
```

**Nix:**

```console
$ rm -rf cbtn
$ time ./image_deid_etl run f90cd746-10017fbb-44472a28-391f1a52-3186cf1a
. . .

real	1m8.341s
user	0m53.190s
sys	0m5.765s
```

</details>

My connection to Orthanc seemed erratic, varying between 500 KB/s and 8 MB/s down when tested natively and from Docker. My home internet connection itself is capable of around ~75 MB/s down. To eliminate any potential noise caused by my network connection, I made sure to cache the study ZIP to only measure performance after the download. 

After ruling out network I/O, there is around a 15% (~10 second) performance penalty from Docker. However, please remember that I only ran one round of these tests, and these results are only from a particular study.

## Testing Instructions

First, ensure [Docker for Mac](https://docs.docker.com/desktop/mac/install/) is installed and current:

```console
$ brew install --cask docker
. . .
$ docker --version   
Docker version 20.10.14, bu ild a224086
$ docker-compose --version
docker-compose version 1.29.2, build 5becea4c
```

And, ensure that your AWS login is current:

```console
$ aws sso login --profile chopd3bprod
```

Next, run `update` to build the container image and initialize the PostgreSQL database:

```console
$ ./scripts/update
```

**Note:** The first time you run `update` will take a while because we have to pull in a bunch of dependencies and compile dcm2niix from scratch. Most of this will be cached, such that subsequent runs are a lot faster.

Then, run `test` to ensure basic ETL functionality works:

```console
$ ./scripts/test
```

**Optional Testing Instructions:**

The introduction of Docker changes the way we configure PyCharm to work with the project. 

First, configure a remote interpreter using Docker Compose:

<details>

![image](https://user-images.githubusercontent.com/1774125/165782145-83fb7b30-19b0-453c-b361-b7f12c90847d.png)

</details>

Next, add a run/debug configuration, taking care to mimic my **Module name** and **Working directory** settings:

<details>

![image](https://user-images.githubusercontent.com/1774125/165786938-46df3484-dd13-4150-a595-5435614588ba.png)

</details>

Now, add a breakpoint to the `check` function, and debug the configuration we just added:

<details>

![image](https://user-images.githubusercontent.com/1774125/165786752-6e9a3ed6-0013-44cd-9f31-eacd733b3e44.png)

</details>
